### DISCUSSION TITLE
Treesitter Concealing

### DISCUSSION THREAD

**Discussion by jgollenz** [nvim-orgmode/orgmode#401] [General]
@kristijanhusak could you please tell me if I understand correctly:

With [this](https://github.com/neovim/neovim/pull/17776) we *could* do link concealing via treesitter, but the org-grammar does not detect links, so we are out of luck anyway?

**Comment by kristijanhusak** [comment]
We are using treesitter concealing for some time now: https://github.com/nvim-orgmode/orgmode/commit/156d8c5bf6f565d1af94ce8f583300467bd76499

Or are you talking about something else?

**Reply by jgollenz** [reply]
I saw the [python example](https://gist.github.com/graelo/2d1f5bc6775432a5c8eb6d6c8b624011) mentioned in the nvim issue and was wondering whether we could do the same in the .scm files. I don't see any `!set conceal` in there, so I thought this might be because unlike the python node `@keyword.operator` we don't have a node `@org.link`, because as far as I know there is no parsing of links happening in the grammar. So we do some extra lifting which can be seen in the commit you mentioned. Is that about right? I'm still not very familiar with treesitter unfortunately.

I stumbled upon this while investigating about virtual text and extmarks in nvim. As this is a general discussion, we might as well talk about the issue I have: concealed links in headlines mess with the tag column. They are left-indented the amount of concealed characters. My first instinct was to inject that amount of spaces in between, but that would actually persist and mess with the tag column when concealing is turned off. So I thought: can we just "pretend" to insert the spaces for padding? Do you know if virtual text can do something like this? Help for `nvim_buf_set_extmark` says: 

```
virt_text_pos : position of virtual text.
Possible values:
  • "eol": right after eol character (default)
  • "overlay": display over the specified
     column, without shifting the underlying
     text.
  • "right_align": display right aligned in the
     window.
```

Neither of those seems to do what I want to achieve.

**Reply by kristijanhusak** [reply]
We are not using built in `conceal` predicate, but our own: https://github.com/nvim-orgmode/orgmode/blob/master/queries/org/markup.scm#L14-L15,  and this is where concealing happens: https://github.com/nvim-orgmode/orgmode/blob/master/lua/orgmode/colors/markup_highlighter.lua#L292-L302

I'm not sure how we could tackle that issue, since concealing reduces the width of the content.
And since we would need to do this only in the headline, it would be tricky.

**Reply by jgollenz** [reply]
I see. Thanks for clarification :+1: 

Somewhere I have a poc of just inserting whitespace after the headline to equal out the width, but like I said, this will then mess with the width when no concealing is happening (e.g. because it is turned off or the file is not viewed in neovim). No good. In theory, we could go through the headline content and count all concealed characters with `synconcealed`, which gives the width we need to equal out. But we cannot touch the actual 'source' line, because `conceal` also doesn't do that. Some sort of virtual text that interacts with the line 'as if' if were really inserted into the buffer would be ideal in my opinion.

Could you elaborate on what the tricky part is regarding only doing it on a headline?

**Reply by kristijanhusak** [reply]
> Could you elaborate on what the tricky part is regarding only doing it on a headline?

Actually this might not be so tricky as I thought. We have the line number of the link, so we can just check if the line starts with asterisk, somewhere around [here](https://github.com/nvim-orgmode/orgmode/blob/master/lua/orgmode/colors/markup_highlighter.lua#L281).
I'm assuming this issue happens also for any markup that is concealed, not only links, since headlines can also have italic or underline markup.
I never used extmarks to do anything with virtual lines, but I guess you can give it a try. You an figure out the length of the concealed text in the code I linked above, depending on if it's a link with or without alias, or a markup.

**Reply by jgollenz** [reply]
> this issue happens also for any markup that is concealed

Yes, strikethrough was what got me into investigating it this time. That's why I think `synconcealed` would be great, as it enables us from treating different cases of concealing the same, just count the concealed chars. Thinking about virtual text: perhaps we could abuse the `overlay` mode for this. Instead of just 'inserting' some spaces into the line to balance concealing out, we can just overlay the tags part of the headline (tags + actual spaces) with virtual text (tags + actual spaces + virtual spaces). 

Won't start on this now though, I'm a bit busy atm

**Reply by jgollenz** [reply]
[398689616](https://github.com/neovim/neovim/pull/9496#issue-398689616) could be interesting here
